### 1. Đọc bộ dữ liệu, cho biết số dòng, số cột và kiểu dữ liệu của các thuộc tính

In [1]:
import pandas as pd
import numpy as np
df = pd.read_csv('OnlineRetail.csv', encoding = 'ISO-8859-1')
df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


+ Bộ dữ liệu có 541909 dòng, 8 cột
+ Thang đo định danh: InvoiceNo, StockCode, Description, InvoiceDate, CustomerID, Country
+ Thang đo tỉ lệ: Quantity, UnitPrice

### 2. Tạo cột mới có tên quý - 'Precious' nhận giá trị 1 nếu ngày lập hoá đơn nằm trong các tháng 1,2,3; nhận giá trị 2 nếu ngày lập hoá đơn nằm trong các tháng 4,5,6; nhận giá trị 3 nếu ngày lập hoá đơn nằm trong các tháng 7,8,9; nhận giá trị 4 nếu ngày lập hoá đơn nằm trong các tháng 10,11,12;

In [3]:
# lọc tháng
a = [x[3:4] for x in df['InvoiceDate']]
df['thang']= a
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country thang  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom     1  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom     1  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom     1  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom     1  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom     1

In [4]:
dk = [(df['thang']=='1')|(df['thang']=='2')|(df['thang']=='3'),(df['thang']=='4')|(df['thang']=='5')|(df['thang']=='6'),(df['thang']=='7')|(df['thang']=='8')|(df['thang']=='9'),(df['thang']=='10')|(df['thang']=='11')|(df['thang']=='12')]
quy = ['1','2','3','4']
df['Precious'] = np.select(dk,quy)
df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country thang Precious  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom     1        1  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom     1        1  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom     1        1  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom     1        1  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom     1        1  
...                 ...        ...         ...             ...   ...      ...  
541904  12/9/2011 12:50       0.85     12680.0          France     9        3  
541905  12/9/2011 12:50       2.10     12680.0          France     9        3  
541906  12/9/2011 12:50       4.15     12680.0          France     9        3  
541907  12/9/2011 12:50       4.15     12680.0          France     9        3  
541908  12/9/2011 12:50       4.95     12680.0          France     9        3  

[541909 rows x 10 columns]

### 3. Tạo một cột mới có tên 'Amount' có giá trị bằng Quantity * UnitPrice

In [5]:
df['Amount'] = df['Quantity']*df['UnitPrice']
df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country thang Precious  \
0        12/1/2010 8:26       2.55     17850.0  United Kingdom     1        1   
1        12/1/2010 8:26       3.39     17850.0  United Kingdom     1        1   
2        12/1/2010 8:26       2.75     17850.0  United Kingdom     1        1   
3        12/1/2010 8:26       3.39     17850.0  United Kingdom     1        1   
4        12/1/2010 8:26       3.39     17850.0  United Kingdom     1        1   
...                 ...        ...         ...             ...   ...      ...   
541904  12/9/2011 12:50       0.85     12680.0          France     9        3   
541905  12/9/2011 12:50       2.10     12680.0          France     9        3   
541906  12/9/2011 12:50       4.15     12680.0          France     9        3   
541907  12/9/2011 12:50       4.15     12680.0          France     9        3   
541908  12/9/2011 12:50       4.95     12680.0          France     9        3   

        Amount  
0        15.30  
1        20.34  
2        22.00  
3        20.34  
4        20.34  
...        ...  
541904   10.20  
541905   12.60  
541906   16.60  
541907   16.60  
541908   14.85  

[541909 rows x 11 columns]

### 4. Tạo một cột mới 'Discount' nhận giá trị 10% nếu Country là 'United Kingdom' và thuộc quý 4, 5% nếu là 'France' ngược lại là 0%

In [6]:
dk2 = [(df['Country']== 'United Kingdom') & (df['Precious']== '4'),df['Country']=='France']
giamgia = [10,5]
df['Discount'] = np.select(dk2,giamgia, default = 0)
df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country thang Precious  \
0        12/1/2010 8:26       2.55     17850.0  United Kingdom     1        1   
1        12/1/2010 8:26       3.39     17850.0  United Kingdom     1        1   
2        12/1/2010 8:26       2.75     17850.0  United Kingdom     1        1   
3        12/1/2010 8:26       3.39     17850.0  United Kingdom     1        1   
4        12/1/2010 8:26       3.39     17850.0  United Kingdom     1        1   
...                 ...        ...         ...             ...   ...      ...   
541904  12/9/2011 12:50       0.85     12680.0          France     9        3   
541905  12/9/2011 12:50       2.10     12680.0          France     9        3   
541906  12/9/2011 12:50       4.15     12680.0          France     9        3   
541907  12/9/2011 12:50       4.15     12680.0          France     9        3   
541908  12/9/2011 12:50       4.95     12680.0          France     9        3   

        Amount  Discount  
0        15.30         0  
1        20.34         0  
2        22.00         0  
3        20.34         0  
4        20.34         0  
...        ...       ...  
541904   10.20         5  
541905   12.60         5  
541906   16.60         5  
541907   16.60         5  
541908   14.85         5  

[541909 rows x 12 columns]

### 5. Tạo cột mới 'Total' nhận giá trị bằng: Amount - Discount

In [8]:
df['Total'] = df['Amount']*(1-df['Discount']/100)
df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country thang Precious  \
0        12/1/2010 8:26       2.55     17850.0  United Kingdom     1        1   
1        12/1/2010 8:26       3.39     17850.0  United Kingdom     1        1   
2        12/1/2010 8:26       2.75     17850.0  United Kingdom     1        1   
3        12/1/2010 8:26       3.39     17850.0  United Kingdom     1        1   
4        12/1/2010 8:26       3.39     17850.0  United Kingdom     1        1   
...                 ...        ...         ...             ...   ...      ...   
541904  12/9/2011 12:50       0.85     12680.0          France     9        3   
541905  12/9/2011 12:50       2.10     12680.0          France     9        3   
541906  12/9/2011 12:50       4.15     12680.0          France     9        3   
541907  12/9/2011 12:50       4.15     12680.0          France     9        3   
541908  12/9/2011 12:50       4.95     12680.0          France     9        3   

        Amount  Discount    Total  
0        15.30         0  15.3000  
1        20.34         0  20.3400  
2        22.00         0  22.0000  
3        20.34         0  20.3400  
4        20.34         0  20.3400  
...        ...       ...      ...  
541904   10.20         5   9.6900  
541905   12.60         5  11.9700  
541906   16.60         5  15.7700  
541907   16.60         5  15.7700  
541908   14.85         5  14.1075  

[541909 rows x 13 columns]